### PR TITLE
Add Iteration related Events

### PR DESCRIPTION
**Pull request by b-chu**
# What does this PR do?
Adds ITERATION_START, ITERATION_END, and ITERATION_CHECKPOINT to Event, Engine, and Callback. Increments iteration during training based on iteration length defined in State. Iteration length is a private variable in State and should have no effect by default.

commit-id:bdbe33f2

---

**Stack**:
- #3087
- #3076 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by mvpatel2000 — COMMENTED** [review]
Can you please add some tests that run multiple iterations? Ideally we have a test that also tries changing iteration length and tests new API in state

**Review by mvpatel2000 — APPROVED** [review]

File: composer/core/engine.py [hunk 1]
```diff
@@ -353,7 +353,11 @@ def _assert_dataloader_and_duration_set(state: State, event: Event):
 
         # dataloader should be set on all events except INIT/BEFORE_LOAD/AFTER_LOAD/EVAL_STANDALONE_START/EVAL_STANDALONE_END
         if event not in {
-                Event.INIT, Event.BEFORE_LOAD, Event.AFTER_LOAD, Event.EVAL_STANDALONE_START, Event.EVAL_STANDALONE_END
+                Event.INIT,
```
Comment (bcui19) [inline composer/core/engine.py:356]: Just want to confirm these are linting chages in these files?

File: composer/core/state.py [hunk 2]
```diff
@@ -606,6 +608,22 @@ def get_elapsed_duration(self) -> Optional[Time[float]]:
             return None
         return self.timestamp.get(self.max_duration.unit) / self.max_duration
 
+    @property
+    def _iteration_length(self):
+        """The length of an iteration."""
+        return self.__iteration_length
+
+    @_iteration_length.setter
+    def _iteration_length(self, iteration_length: Optional[Union[str, Time[int]]]):
```
Comment (bcui19) [inline composer/core/state.py:617]: Maybe leave some detail as to what an iteration is/why it needs to have time units of an epoch?

File: composer/trainer/trainer.py [hunk 3]
```diff
@@ -2244,6 +2247,14 @@ def _train_loop(self) -> None:
 
                 self.engine.run_event(Event.EPOCH_CHECKPOINT)
 
+                # Increment iteration
+                if (self.state._iteration_length is not None and
+                        self.state.timestamp.epoch_in_iteration == self.state._iteration_length):
+                    self.state.previous_timestamp = self.state.timestamp
+                    self.state.timestamp = self.state.timestamp.to_next_iteration()
```
Comment (bcui19) [inline composer/trainer/trainer.py:2254]: Where is `to_next_iteration()` implemented? If it a part of this PR or am I missing sth?

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -1138,6 +1138,25 @@ def test_compile_uncompile_model_weights_trainer_fit(
...
+        trainer = Trainer(
+            model=model,
+            max_duration='5ep',
+            train_dataloader=train_dataloader,
+        )
+        trainer.state._iteration_length = '2ep'
+        trainer.fit()
+
+        assert trainer.state.timestamp.epoch == Time(5, TimeUnit.EPOCH)
+        assert trainer.state.timestamp.iteration == Time(2, TimeUnit.ITERATION)
```
Comment (bcui19) [inline tests/trainer/test_trainer.py:1158]: Just for my sanity, because we're in the middle of the second iteration, this is why the assert passes?

**Comment by bcui19** [comment]
From my perspective, this seems good. Added some comments, but yea it'd be good to add more testing/also include tests on if we have early breaks from an epoch in an iteration.

File: composer/core/engine.py (hunk 1 above)
Comment (b-chu) [reply composer/core/engine.py:356]: Yes

File: composer/trainer/trainer.py (hunk 3 above)
Comment (b-chu) [reply composer/trainer/trainer.py:2254]: It got added in #3017

**Comment by b-chu** [comment]
For early breaks in iteration, this would be implemented by users via callback so wouldn't be tested here

**Pull request merged**